File: scripts/autotiling.py
#!/usr/bin/env python3
"""
autotiling.py — Fibonacci spiral autotiling for Sway.

Listens to Sway IPC window events and automatically sets split direction
(splith / splitv) on the focused window based on its aspect ratio (width vs height).
This creates a continuous Fibonacci spiral tiling layout for new windows.
"""

import json
import os
import socket
import struct
import subprocess
import sys

MAGIC = b"i3-ipc"
RUN_COMMAND = 0
GET_TREE = 4
SUBSCRIBE = 2


def get_socket_path():
    sock = os.environ.get("SWAYSOCK")
    if not sock:
        try:
            res = subprocess.check_output(["swaymsg", "-t", "get_version"], text=True)
            sock = os.environ.get("SWAYSOCK")
        except Exception:
            pass
    return sock


def pack_message(msg_type, payload=""):
    payload_bytes = payload.encode("utf-8")
    header = struct.pack("=6sII", MAGIC, len(payload_bytes), msg_type)
    return header + payload_bytes


def unpack_header(sock):
    header = sock.recv(14)
    if len(header) < 14:
        return None, None
    magic, length, msg_type = struct.unpack("=6sII", header)
    if magic != MAGIC:
        return None, None
    body = b""
    while len(body) < length:
        chunk = sock.recv(length - len(body))
        if not chunk:
            break
        body += chunk
    return msg_type, body.decode("utf-8")


def find_focused(node):
    if node.get("focused"):
        return node
    for child in node.get("nodes", []) + node.get("floating_nodes", []):
        res = find_focused(child)
        if res:
            return res
    return None


def get_tree(sock):
    sock.sendall(pack_message(GET_TREE))
    msg_type, payload = unpack_header(sock)
    if payload:
        return json.loads(payload)
    return None


def send_command(sock, cmd):
    sock.sendall(pack_message(RUN_COMMAND, cmd))
    unpack_header(sock)


def main():
    sock_path = get_socket_path()
    if not sock_path or not os.path.exists(sock_path):
        sys.exit("SWAYSOCK not available")

    cmd_sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    cmd_sock.connect(sock_path)

    event_sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    event_sock.connect(sock_path)

    event_sock.sendall(pack_message(SUBSCRIBE, json.dumps(["window"])))
    unpack_header(event_sock)

    while True:
        msg_type, payload = unpack_header(event_sock)
        if payload is None:
            break
        try:
            event = json.loads(payload)
        except Exception:
            continue

        change = event.get("change")
        if change in ("focus", "new"):
            tree = get_tree(cmd_sock)
            if not tree:
                continue
            focused = find_focused(tree)
            if not focused:
                continue

            rect = focused.get("rect", {})
            width = rect.get("width", 0)
            height = rect.get("height", 0)

            if width > 0 and height > 0:
                if width >= height:
                    send_command(cmd_sock, "splith")
                else:
                    send_command(cmd_sock, "splitv")


if __name__ == "__main__":
    main()

File: rust/cumulus-dotfiles/src/autotiling.rs
//! `cumulus-autotiling` — Fibonacci spiral autotiling for Sway.
//! Ports `scripts/autotiling.py` to std Rust.

use crate::context::Context;
use crate::error::{Error, Result};
use std::env;
use std::io::{Read, Write};
use std::os::unix::net::UnixStream;
use std::path::Path;
use std::process::Command;

const MAGIC: &[u8; 6] = b"i3-ipc";
const RUN_COMMAND: u32 = 0;
const SUBSCRIBE: u32 = 2;
const GET_TREE: u32 = 4;

fn get_socket_path() -> Option<String> {
    if let Ok(sock) = env::var("SWAYSOCK") {
        if !sock.is_empty() && Path::new(&sock).exists() {
            return Some(sock);
        }
    }
    if let Ok(out) = Command::new("swaymsg").args(["-t", "get_version"]).output() {
        if out.status.success() {
            if let Ok(sock) = env::var("SWAYSOCK") {
                if !sock.is_empty() && Path::new(&sock).exists() {
                    return Some(sock);
                }
            }
        }
    }
    None
}

fn pack_message(msg_type: u32, payload: &str) -> Vec<u8> {
    let payload_bytes = payload.as_bytes();
    let len = payload_bytes.len() as u32;
    let mut buf = Vec::with_capacity(14 + payload_bytes.len());
    buf.extend_from_slice(MAGIC);
    buf.extend_from_slice(&len.to_le_bytes());
    buf.extend_from_slice(&msg_type.to_le_bytes());
    buf.extend_from_slice(payload_bytes);
    buf
}

fn unpack_header(stream: &mut UnixStream) -> std::io::Result<(u32, String)> {
    let mut header = [0u8; 14];
    stream.read_exact(&mut header)?;
    if &header[0..6] != MAGIC {
        return Err(std::io::Error::new(
            std::io::ErrorKind::InvalidData,
            "Invalid IPC magic header",
        ));
    }
    let length = u32::from_le_bytes(header[6..10].try_into().unwrap());
    let msg_type = u32::from_le_bytes(header[10..14].try_into().unwrap());

    let mut body = vec![0u8; length as usize];
    stream.read_exact(&mut body)?;
    let payload = String::from_utf8_lossy(&body).into_owned();
    Ok((msg_type, payload))
}

fn send_ipc(stream: &mut UnixStream, msg_type: u32, payload: &str) -> std::io::Result<String> {
    let buf = pack_message(msg_type, payload);
    stream.write_all(&buf)?;
    stream.flush()?;
    let (_, resp) = unpack_header(stream)?;
    Ok(resp)
}

#[derive(Debug, PartialEq, Eq)]
struct Rect {
    width: u64,
    height: u64,
}

fn find_focused_rect(json_str: &str) -> Option<Rect> {
    let focused_idx = json_str
        .find("\"focused\": true")
        .or_else(|| json_str.find("\"focused\":true"))?;

    let rect_idx = json_str[focused_idx..].find("\"rect\"")?;
    let rect_substr = &json_str[focused_idx + rect_idx..];

    let width_val = extract_json_number(rect_substr, "width")?;
    let height_val = extract_json_number(rect_substr, "height")?;

    Some(Rect {
        width: width_val,
        height: height_val,
    })
}

fn extract_json_number(s: &str, key: &str) -> Option<u64> {
    let pattern = format!("\"{key}\"");
    let key_idx = s.find(&pattern)?;
    let after_key = &s[key_idx + pattern.len()..];
    let colon_idx = after_key.find(':')?;
    let num_str = after_key[colon_idx + 1..]
        .trim_start()
        .chars()
        .take_while(|c| c.is_ascii_digit())
        .collect::<String>();
    num_str.parse::<u64>().ok()
}

pub fn run(_ctx: &Context, args: &[String]) -> Result<()> {
    if args.iter().any(|a| a == "-h" || a == "--help") {
        println!("cumulus-autotiling — Fibonacci spiral autotiling for Sway.");
        println!("Listens to Sway IPC window events and sets split direction automatically.");
        return Ok(());
    }

    let sock_path = get_socket_path().ok_or_else(|| Error::new("SWAYSOCK not available"))?;

    let mut cmd_sock = UnixStream::connect(&sock_path)
        .map_err(|e| Error::new(format!("failed to connect to SWAYSOCK: {e}")))?;
    let mut event_sock = UnixStream::connect(&sock_path)
        .map_err(|e| Error::new(format!("failed to connect to SWAYSOCK: {e}")))?;

    send_ipc(&mut event_sock, SUBSCRIBE, "[\"window\"]")
        .map_err(|e| Error::new(format!("failed to subscribe to sway events: {e}")))?;

    while let Ok((_, payload)) = unpack_header(&mut event_sock) {
        let is_focus =
            payload.contains("\"change\":\"focus\"") || payload.contains("\"change\": \"focus\"");
        let is_new =
            payload.contains("\"change\":\"new\"") || payload.contains("\"change\": \"new\"");

        if is_focus || is_new {
            if let Ok(tree_payload) = send_ipc(&mut cmd_sock, GET_TREE, "") {
                if let Some(rect) = find_focused_rect(&tree_payload) {
                    if rect.width > 0 && rect.height > 0 {
                        let cmd = if rect.width >= rect.height {
                            "splith"
                        } else {
                            "splitv"
                        };
                        let _ = send_ipc(&mut cmd_sock, RUN_COMMAND, cmd);
                    }
                }
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_focused_rect() {
        let sample = r#"{
            "id": 1,
            "focused": false,
            "rect": { "width": 3840, "height": 1080 },
            "nodes": [
                {
                    "id": 2,
                    "focused": true,
                    "rect": { "x": 0, "y": 0, "width": 1920, "height": 1080 }
                }
            ]
        }"#;
        let rect = find_focused_rect(sample).unwrap();
        assert_eq!(rect.width, 1920);
        assert_eq!(rect.height, 1080);
    }

    #[test]
    fn test_extract_json_number() {
        let snippet = r#""rect": { "x": 100, "y": 200, "width": 1920, "height": 1080 }"#;
        assert_eq!(extract_json_number(snippet, "width"), Some(1920));
        assert_eq!(extract_json_number(snippet, "height"), Some(1080));
    }
}
